### PR TITLE
Support for boolean and wrapped attributes

### LINKED ISSUE #11: Support boolean attributes
input(type="password" required)

should become <input type="password" required>

### PR DESCRIPTION
**Pull request by doomspork**
This PR adds support for wrapped attributes and boolean attributes: `input (type="text" required)`.  @Rakoth if you have a moment, I'd welcome your feedback.

This closes #11.

### CODE REVIEW

File: lib/slim_fast/parser.ex [hunk 1]
```diff
@@ -95,6 +71,41 @@ defmodule SlimFast.Parser do
     end
   end
 
+  @attr_regex ~r/([\w-]+)\s*=\s*(".+"|[\w-]+)/
+  defp parse_attributes(line, acc \\ [])
+  defp parse_attributes("", acc), do: {"", acc}
+  defp parse_attributes(nil, acc), do: {"", acc}
+  defp parse_attributes("(" <> line, acc), do: unwrap_attributes(line)
```
Comment (Rakoth) [inline lib/slim_fast/parser.ex]: I would suggest to split this function into 2: `parse_attributes/1`, which handles nil and deals with attributes wrapping and after that calls `parse_attributes/2` which parses attributes and accumulate the result.
Now it is unclear why `acc` is not passed in nested call.

Comment (doomspork) [inline lib/slim_fast/parser.ex]: I'll look into incorporating these changes, thanks for the feedback :+1:

File: lib/slim_fast/parser.ex [hunk 2]
```diff
@@ -95,6 +71,40 @@ defmodule SlimFast.Parser do
     end
   end
 
+  @attr_regex ~r/([\w-]+)\s*=\s*(".+"|[\w-]+)/
+  defp parse_attributes(""), do: {"", []}
```
Comment (doomspork) [inline lib/slim_fast/parser.ex:114]: @Rakoth is this more inline with what you had in mind?  Thanks again

Comment (Rakoth) [inline lib/slim_fast/parser.ex:114]: Yes, looks great!

File: lib/slim_fast/parser.ex [hunk 3]
```diff
@@ -95,6 +71,40 @@ defmodule SlimFast.Parser do
     end
   end
 
+  @attr_regex ~r/([\w-]+)\s*=\s*(".+"|[\w-]+)/
```
Comment (Rakoth) [inline lib/slim_fast/parser.ex]: It does not support expression like `meta content=@user.name` and other simple elixir code =(

Comment (doomspork) [inline lib/slim_fast/parser.ex]: So `@` and `#{}`, right?  Are there others?

Comment (doomspork) [inline lib/slim_fast/parser.ex]: Probably should include single quotes too.  Something like this perhaps: `\s*(["'].+['"]|\#{.+}|[@\w.-]+)`

Comment (Rakoth) [inline lib/slim_fast/parser.ex]: For starters `@user.name`, `user.name` and `user["name"]` support will be great.
Ideally, any code without spaces, leaving `attr=(code with spaces)` case for the future.

Comment (doomspork) [inline lib/slim_fast/parser.ex]: Sounds, good.  I'll take a stab at that.  Thanks :+1:

File: test/parser_test.exs [hunk 4]
```diff
@@ -19,26 +19,61 @@ defmodule ParserTest do
...
+    assert opts[:attributes] == [required: {:eex, content: "true", inline: true}, type: "text"]
+  end
+
   test "parses attributes with interpolation" do
     {_, {:meta, opts}} = ~S(meta content="one#{two}") |> Parser.parse_line
 
     assert opts[:attributes] == [content: {:eex, content: ~S("one#{two}"), inline: true}]
   end
 
+  test "parses attributes with elixir code" do
```
Comment (doomspork) [inline test/parser_test.exs:58]: @Rakoth I updated the regex to accommodate attributes with elixir code.  I used your three examples as tests.

**Comment by doomspork** [comment]
@Rakoth once I get your :+1: on these changes I'm going to release `0.5.0` with your EEx macros, comment support, wrapped attributes, and boolean attributes.

File: lib/slim_fast/parser.ex [hunk 5]
```diff
@@ -95,6 +75,39 @@ defmodule SlimFast.Parser do
...
+        parse_attributes(line, acc)
+      nil -> {line, acc}
+    end
+  end
+
+  defp unwrap_attributes(line) do
+    line = case Regex.run(@boolean_regex, line) do
+             [_|[match|_]] ->
+               match = String.strip(match)
+               String.replace(line, match, match <> "=true")
```
Comment (Rakoth) [inline lib/slim_fast/parser.ex]: This would replace all inclusions of matched string, so the following assert fails

``` elixir
assert render(~s(div [ab="ab" a] a)) == ~s(<div ab="ab" a>a</span>)
```

**Comment by Rakoth** [comment]
Unfortunately there is some error with regex, or somewhere else, so these assets fails:

``` elixir
assert render(~s(div [a="a" b] c)) == ~s(<div a="a" b>c</div>)
assert render(~s(div [a b="b"] c)) == ~s(<div a b="b">c</div>)
```

File: lib/slim_fast/parser.ex (hunk 5 above)
Comment (doomspork) [inline lib/slim_fast/parser.ex]: Haha, clever.  Good catch!

**Comment by doomspork** [comment]
@Rakoth is `assert render(~s(div [a b="b"] c)) == ~s(<div a b="b">c</div>)` supported?  I looked through the Slim documentation and at first glance it appears that the boolean attribute has to be the last attribute in the list.

**Comment by Rakoth** [comment]
It is supported, also there can be more than one boolean attribute like in 

``` slim
script[defer async src="..."]
```

**Comment by doomspork** [comment]
Alright, good to know.  I'm going to need to re-think my approach for the attributes.

**Comment by Rakoth** [comment]
I think of algorithm such as this:
1) get rid of inline childs
2) remove wrapping brackets
3) remove any extra spaces (around `=` and sequences of more than one space between attributes)
4) split by an space
5) deal with each attribute, handle case when there is no equal sign

**Comment by doomspork** [comment]
How would you accomplish getting rid of the inline children?  Right now they're what's left over after the attributes have been yanked.

The one step I was unsure of is "3) remove any extra spaces (around = and sequences of more than one space between attributes)".  The side effect is removing whitespace from within `" "` strings that contain an `=` but it may just be an edge case.

**Comment by Rakoth** [comment]
Yeah, you right, it will not handle "-quoted strings with spaces.

File: test/renderer_test.exs [hunk 6]
```diff
@@ -82,4 +82,10 @@ defmodule RendererTest do
 
     assert render(slim) == ~s(<p class="test"></p>)
   end
+
+  test "render tag with boolean attribute" do
+    assert render(~s(div [ab="ab" a] a)) == ~s(<div ab="ab" a>a</div>)
```
Comment (doomspork) [inline test/renderer_test.exs:211]: @Rakoth I added the examples you gave me in the comments.

**Comment by doomspork** [comment]
Alright @Rakoth, I think I'm ready for you to take another peek at this.

File: lib/slim_fast/parser.ex [hunk 7]
```diff
@@ -105,6 +90,49 @@ defmodule SlimFast.Parser do
...
+    parts = String.split(head, ~r/=/, parts: 2)
+    attr = case parts do
+             [key, value] -> html_attribute(key, value)
+             [key] -> html_attribute(key, "true")
+             _ -> []
+           end
+    parse_attributes(tail, attr ++ acc)
+  end
+
+  defp parse_wrapped_attributes(line) do
```
Comment (doomspork) [inline lib/slim_fast/parser.ex]: I should move this so it's in :abc: order.

File: lib/slim_fast/parser.ex [hunk 8]
```diff
@@ -105,13 +90,47 @@ defmodule SlimFast.Parser do
     end
   end
 
-  defp parse_line(@blank, _line), do: @blank
-  defp parse_line(@content, line), do: line |> String.slice(1..-1) |> String.strip |> parse_eex_string
-  defp parse_line(@comment, line), do: line |> String.slice(1..-1) |> parse_comment
-  defp parse_line(@html, line), do: line |> String.strip |> parse_eex_string
+  defp parse_attributes(""), do: {"", []}
```
Comment (doomspork) [inline lib/slim_fast/parser.ex:114]: @Rakoth are you aware if there is any benefit to updating the code to use the module attribute `@blank` for blank strings like such: 

``` elixir
defp parse_attributes(@blank), do: {@blank, []}
```

**Comment by doomspork** [comment]
Rebased against the latest master.

File: lib/slim_fast/compiler.ex [hunk 9]
```diff
@@ -9,15 +9,22 @@ defmodule SlimFast.Compiler do
 
   defp render_attribute(_, []), do: ""
   defp render_attribute(_, ""), do: ""
+  defp render_attribute(name, {:eex, opts}) do
+    value = opts[:content]
+    case value do
+      "true"  -> name
+      "false" -> ""
+      "nil"   -> ""
+      _       -> "#{name}=<%=#{value}%>"
```
Comment (Rakoth) [inline lib/slim_fast/compiler.ex]: We have to wrap `<%=#{value}%>` here into `""`, so attributes values with spaces do not produce extra attributes

``` elixir
assert render(~S(div ab="#{b} a" a), b: "b") == ~s(<div ab="b a">a</div>)
```

File: lib/slim_fast/parser.ex (hunk 8 above)
Comment (Rakoth) [inline lib/slim_fast/parser.ex:114]: I don't really see any benefits here, especially in `do:` part

**Comment by Rakoth** [comment]
I have one more failing scenario, but I don't yet figured out why is it

``` elixir
assert render(~S(div [ab="a #{b}" a] a), b: "b") == ~s(<div ab="a b" a>a</div>)
```

File: lib/slim_fast/compiler.ex (hunk 9 above)
Comment (doomspork) [inline lib/slim_fast/compiler.ex]: Good catch.  Unfortunately, I don't think it's as easy as `"#{name}=\"<%=#{value}%>\""`; EEx can return non-string values and HTML supports non-quoted attributes: https://html.spec.whatwg.org/multipage/syntax.html#syntax-attributes

I can see that your `parse_eex_string/1` function has been called and appropriately escaped everything, transforming `"\#{b} a"`to `"\"\#{b} a\""`.  I'll continue to dig in.

File: test/renderer_test.exs (hunk 6 above)
Comment (Rakoth) [inline test/renderer_test.exs:211]: As far as I know there is no difference between unquoted attribute and quoted attribute with the same value. Also there is no support for unquoted attributes in slim. If somebody for some reason wants unquoted attribute in result html, he can use inline html for that.

Comment (doomspork) [inline test/renderer_test.exs:211]: Thanks for the insight :+1:

File: lib/slim_fast/compiler.ex (hunk 9 above)
Comment (Rakoth) [inline lib/slim_fast/compiler.ex]: I think it is good enough for now. There is no unquoted attributes support in slim, except you can use inline html for that.

File: test/renderer_test.exs [hunk 10]
```diff
@@ -169,4 +169,11 @@ defmodule RendererTest do
     And no spaces</p>
     """, ?\n)
   end
+
+  test "render tag with boolean attribute" do
+    assert render(~s(div [ab="ab" a] a)) == ~s(<div ab="ab" a>a</div>)
+    assert render(~s(div [a b="b"] c)) == ~s(<div a b="b">c</div>)
+    assert render(~S(div ab="#{b} a" a), b: "b") == ~s(<div ab="b a">a</div>)
```
Comment (doomspork) [inline test/renderer_test.exs:213]: @Rakoth your latest case has been added to the tests

File: lib/slim_fast/compiler.ex [hunk 11]
```diff
@@ -9,15 +9,22 @@ defmodule SlimFast.Compiler do
 
   defp render_attribute(_, []), do: ""
   defp render_attribute(_, ""), do: ""
+  defp render_attribute(name, {:eex, opts}) do
+    value = opts[:content]
+    case value do
+      "true"  -> name
+      "false" -> ""
+      "nil"   -> ""
+      _       -> "#{name}=\"<%=#{value}%>\""
```
Comment (doomspork) [inline lib/slim_fast/compiler.ex]: Added quotes.

File: test/renderer_test.exs (hunk 10 above)
Comment (Rakoth) [inline test/renderer_test.exs:213]: There is another test case I've added in comment to PR, much like this one, but with attributes wrapping, it is still failing. =(

Comment (doomspork) [inline test/renderer_test.exs:213]: You are the master of finding these, I'll dig in  :+1:

**Comment by doomspork** [comment]
@Rakoth this has been updated to accommodate for the issue you discovered.

**Comment by Rakoth** [comment]
Great to hear :+1:
